src/main/java/com/mcmoddev/bot/helpers/forge/ForgeVersionHelper.java

```diff
@@ -0,0 +1,103 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class ForgeVersionHelper {
+    private static final String VERSION_URL = "https://files.minecraftforge.net/maven/net/minecraftforge/forge/promotions_slim.json";
+    private static final Pattern VERSION_REGEX = Pattern.compile("(.+?)-(.+)");
+
+    private static final Gson gson = new Gson();
+
+    public static String getLatestVersion(List<String> versions) {
+        SemVer latest = new SemVer(versions.get(0));
+
+        for (int i = 0; i < versions.size(); i++) {
+            SemVer ver = new SemVer(versions.get(i));
+            if(latest.compareTo(ver) == -1) {
+                latest = ver;
+            }
+        }
+
+        return latest.toString();
+    }
+
+    public static ForgeVersion getForgeVersionsForMcVersion(String mcVersion) throws Exception {
+        return getForgeVersions().get(mcVersion);
+    }
+
+    public static MinecraftForgeVersion getLatestMcVersionForgeVersions() throws Exception {
+        Map<String, ForgeVersion> versions = getForgeVersions();
+
+        String latest = getLatestVersion(new ArrayList<String>(versions.keySet()));
+
+        return new MinecraftForgeVersion(latest, versions.get(latest));
+    }
+
+    private static InputStreamReader openUrl(String url) throws Exception {
+        URL urlObj = new URL(VERSION_URL);
+
+        return new InputStreamReader(urlObj.openStream());
+    }
+
+    public static Map<String, ForgeVersion> getForgeVersions() throws Exception {
+        InputStreamReader reader = openUrl(VERSION_URL);
+
+        ForgePromoData data = gson.fromJson(reader, ForgePromoData.class);
+
+        // Collect version data
+        Map<String, ForgeVersion> versions = new HashMap<>();
+
+        String currentVersion = "";
+        String latest = "";
+        String recommended = "";
+
+        for (Map.Entry<String,String> entry : data.promos.entrySet()) {
+            String mc = entry.getKey();
+            String forge = entry.getValue();
+
+            VersionMeta meta = getMCVersion(mc);
+
+            if(versions.containsKey(meta.version)) {
+                ForgeVersion version = versions.get(meta.version);
+                if(meta.state.equals("recommended")) {
+                    version.setRecommended(forge);
+                } else {
+                    version.setLatest(forge);
+                }
+            } else {
+                ForgeVersion version = new ForgeVersion();
+                if(meta.state.equals("recommended")) {
+                    version.setRecommended(forge);
+                } else {
+                    version.setLatest(forge);
+                }
+                versions.put(meta.version, version);
+            }
+        }
+
+        // For some reason there's an entry named
+        // latest-1.7.10 from the API, remove it.
+        versions.remove("latest");
```

Comment:
Is this supposed to say latest, or latest-1.7.10 like the comment says?

Reply:
The version is parsed with a regex that matches `(.*)-(.*)` so it matches this random `latest-1.7.10` and parse "latest" as the version and puts it in the `version` map with the key "latest", that's why it's simply "latest".

![image](https://user-images.githubusercontent.com/25696312/73582343-c15c0480-445a-11ea-924d-0efbf1b129fd.png)